x/dualstaking/keeper/delegate.go

```diff
@@ -0,0 +1,563 @@
+	if err != nil {
+		return utils.LavaFormatWarning("failed to decrease delegation", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: from},
+			utils.Attribute{Key: "amount", Value: amount.String()},
+		)
+	}
+
+	err = k.increaseDelegation(ctx, delegator, to, toChainID, amount, nextEpoch)
+	if err != nil {
+		return utils.LavaFormatWarning("failed to increase delegation", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: to},
+			utils.Attribute{Key: "amount", Value: amount.String()},
+		)
+	}
+
+	// no need to transfer funds, because they remain in the dualstaking module
+	// (specifically in types.BondedPoolName).
+
+	return nil
+}
+
+// Unbond lets a delegator get its delegated coins back from a provider. The
+// delegation ends immediately, but coins are held for unstakeHoldBlocks period
+// before released and transferred back to the delegator. The rewards from the
+// provider will be updated accordingly (or terminate) from the next epoch.
+// (effective on next epoch)
+func (k Keeper) Unbond(ctx sdk.Context, delegator, provider, chainID string, amount sdk.Coin) error {
+	nextEpoch, err := k.getNextEpoch(ctx)
+	if err != nil {
+		return err
+	}
+
+	if _, err = sdk.AccAddressFromBech32(delegator); err != nil {
+		return utils.LavaFormatWarning("invalid delegator address", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+		)
+	}
+
+	if _, err = sdk.AccAddressFromBech32(provider); err != nil {
+		return utils.LavaFormatWarning("invalid provider address", err,
+			utils.Attribute{Key: "provider", Value: provider},
+		)
+	}
+
+	if err := validateCoins(amount); err != nil {
+		return err
+	} else if amount.IsZero() {
+		return nil
+	}
+
+	err = k.decreaseDelegation(ctx, delegator, provider, chainID, amount, nextEpoch)
+	if err != nil {
+		return utils.LavaFormatWarning("failed to decrease delegation", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "amount", Value: amount.String()},
+		)
+	}
+
+	// in unbonding the funds to not return immediately to the delegator; instead
+	// they transfer to the NotBondedPoolName module account, where they are held
+	// for the hold period before they are finally released.
+
+	k.bondedTokensToNotBonded(ctx, amount.Amount)
+
+	err = k.setUnbondingTimer(ctx, delegator, provider, chainID, amount)
+	if err != nil {
+		return utils.LavaFormatError("failed to unbond delegated coins", err,
+			utils.Attribute{Key: "amount", Value: amount},
+		)
+	}
+
+	return nil
+}
+
+var space = " "
+
+// encodeForTimer generates timer key unique to a specific delegation; thus it
+// must include the delegator, provider, and chainID.
+func encodeForTimer(delegator, provider, chainID string) []byte {
+	dlen, plen, clen := len(delegator), len(provider), len(chainID)
```

Comment:
can we sanitize the separator here? since we decode from it later

Reply:
unsure what you mean by "sanitize"?
delegator, provider and chainID were all validated by now, and therefore may not include spaces.
see x/dualstaking/types/keys.go line 35 for rationale of choosing "space" as separator.